customcommands/interval_test.go

```diff
@@ -1,315 +1,335 @@
+			}
+			got := CalcNextRunTime(cmd, tt.RefTime).UTC()
+
+			testDesc := fmt.Sprintf("interval every %d mins\n\texclude hours: %v\n\texclude days: %v",
+				tt.Interval, tt.Exclude.Hours, tt.Exclude.Days)
+			testHelper(t, testDesc, got, tt.Want, tt.RefTime)
+		})
 	}
+}
 
-	cc.TextTrigger = "0 * * * 2"
-
-	nextRun = CalcNextRunTime(cc, tim)
-	expected = tim.UTC().Add(time.Hour * 24)
-
-	if nextRun != expected {
-		t.Error("incorrect next run, should be: ", expected, ", got: ", nextRun)
+func TestCronNextRunTime(t *testing.T) {
+	var now = time.Now().UTC()
+	var zero = time.Time{}.UTC()
+
+	tests := []struct {
+		Name    string
+		Cron    string
+		Exclude exclude
+		RefTime time.Time
+
+		Want expectedTime
+	}{
+		{Name: "basic on minute 5",
+			Cron:    "5 * * * *",
+			RefTime: zero.Add(3 * time.Minute),
+			Want:    runOnDate(minute(5))},
+
+		{Name: "on minute 0 of day 2",
+			Cron:    "0 * 2 * *",
+			RefTime: zero,
+			Want:    runOnDate(day(2), hour(0), minute(0))},
+
+		{Name: "on minute 0 of feb",
+			Cron:    "0 * * 2 *",
+			RefTime: zero,
+			Want:    runOnDate(month(time.February), day(1), hour(0), minute(0))},
+
+		{Name: "on minute 0 of tues",
+			Cron:    "0 * * * 2",
+			RefTime: zero,
+			Want:    runOnDate(day(2), hour(0), minute(0))},
+
+		{Name: "exclude all hours",
+			Cron:    "* * * * *",
+			Exclude: exclude{Hours: allHours()},
+			RefTime: now,
+			Want:    neverRuns()},
+
+		{Name: "exclude hour 0 from cron running on hour 0",
+			Cron:    "* 0 * * * *",
+			Exclude: exclude{Hours: []int64{0}},
+			RefTime: now,
+			Want:    neverRuns()},
+
+		{Name: "exclude sunday from cron running on day 0",
+			Cron:    "* * * * 0",
+			Exclude: exclude{Days: []time.Weekday{time.Sunday}},
+			RefTime: now,
+			Want:    neverRuns()},
+
+		{Name: "on minute 5 with exclude hours",
+			Cron:    "5 * * * *",
+			Exclude: exclude{Hours: []int64{0, 1}},
+			RefTime: zero.Add(15 * time.Minute),
+			Want:    runOnDate(hour(2), minute(5))},
+
+		{Name: "on minute 0 on 2nd day with exclude hours",
+			Cron:    "0 * 2 * *",
+			Exclude: exclude{Hours: []int64{0, 1}},
+			RefTime: zero.Add(18 * time.Minute),
+			Want:    runOnDate(day(2), hour(2), minute(0))},
+
+		{Name: "on minute 0 in february with exclude hours",
+			Cron:    "0 * * 2 *",
+			Exclude: exclude{Hours: []int64{0, 1}},
+			RefTime: zero,
+			Want:    runOnDate(month(time.February), day(1), hour(2), minute(0))},
+
+		{Name: "on minute 0 on tues with exclude hours",
+			Cron:    "0 * * * 2",
+			Exclude: exclude{Hours: []int64{0, 1}},
+			RefTime: zero,
+			Want:    runOnDate(day(2), hour(2), minute(0))},
+
+		{Name: "on minute 5 with exclude days",
+			Cron:    "5 * * * *",
+			Exclude: exclude{Days: []time.Weekday{time.Sunday, time.Monday}},
+			RefTime: zero,
+			Want:    runOnDate(day(2), hour(0), minute(5))},
+
+		{Name: "on minute 0 on day 2,3 with exclude days",
+			Cron:    "0 * 2,3 * *",
+			Exclude: exclude{Days: []time.Weekday{time.Sunday, time.Monday}},
```

Comment:
I believe this is supposed to exclude Tuesday with an expected date of Wednesday, I don’t think the exclude days are making a difference here. That might have been a mistake in my original.

Reply:
Your original test was probably correct :-) It's likely I threw in an off-by-one when converting Days to `time.Weekday` values from `int64`s. Will fix.